cmd/kubeadm/app/phases/upgrade/postupgrade.go

```diff
@@ -122,6 +123,18 @@ func WriteKubeletConfigFiles(cfg *kubeadmapi.InitConfiguration, patchesDir strin
 	}
 
 	errs := []error{}
+
+	if features.Enabled(cfg.FeatureGates, features.NodeLocalCRISocket) {
+		containerRuntimeEndpoint, err := kubeletphase.ReadKubeadmFlags(filepath.Join(kubeletDir, kubeadmconstants.KubeletEnvFileName), out)
```

Comment:
I think we have 2 scenarios here.
1. upgrade an old node(no `instance-config.yaml`): you should get from `kubeadm-flags.env`.
2. upgrade a new node: you can use instance-config.yaml.

So here, we should check instance-config.yaml at first and use it if we already have one. If not, try `kubeadm-flags.env`.

Reply:
fair.